spanner/google/cloud/spanner/streamed.py

```diff
@@ -130,7 +159,35 @@ def consume_next(self):
 
         Parse the result set into new/existing rows in :attr:`_rows`
         """
-        response = six.next(self._response_iterator)
+        sleep_generator = retry.exponential_sleep_generator(
+            initial=retry._DEFAULT_INITIAL_DELAY,
+            maximum=retry._DEFAULT_MAXIMUM_DELAY)
+
+        deadline = time.time() + _RESTART_DEADLINE
+
+        response = None
+
+        for sleep in sleep_generator:
+            try:
+                response = six.next(self._response_iterator)
+            except exceptions.ServiceUnavailable:
+
+                if self._resume_token in (None, b''):
+                    raise
+
+                if time.time() > deadline:
+                    raise
+
+                self._response_iterator = self._restart(
+                    resume_token=self._resume_token)
+
+                time.sleep(sleep)
+            else:
+                break
+
+        if response is None:
+            raise ValueError('Sleep generator stopped yielding sleep values.')
```

Comment:
1. You should add a `raises` section?
1. It seems this error message is not comprehensive enough? I.e. `response` could be `None` because `self._response_iterator` broke in some unexpected way

Reply:
> You should add a raises section?

9ae4121

> I.e. response could be None because self._response_iterator broke in some unexpected way.

`self._response_iterator` is our code (not user-supplied), and is mandated to return a protobuf.

Reply:
> `self._response_iterator` is our code (not user-supplied)

I realize. It comes from gRPC or hand-rolled here?

Reply:
it is the GAPIC streaming iterator returned from the call to `SpannerClient.{read_streaming,execute_streaming_sql}`.